sdks/python/apache_beam/transforms/stats.py

```diff
@@ -234,3 +239,419 @@ def extract_output(accumulator):
+    # final output.
+    self.unbuffered_elements = unbuffered_elements
+
+  def is_empty(self):
+    """Check if the buffered & unbuffered elements are empty or not."""
+    return not self.unbuffered_elements and not self.buffers
+
+
+class ApproximateQuantilesCombineFn(CombineFn):
+  """
+  This combiner gives an idea of the distribution of a collection of values
+  using approximate N-tiles. The output of this combiner is the list of size of
+  the number of quantiles (num_quantiles), containing the input values of the
+  minimum value item of the list, the intermediate values (n-tiles) and the
+  maximum value item of the list, in the sort order provided via comparator
+  (compare or key).
+
+  If there are fewer values to combine than the number of quantile
+  (num_quantiles), then the resulting list will contain all the values being
+  combined, in sorted order.
+
+  If no comparator (compare or key) is provided, then the results are sorted
+  in the natural order.
+
+  To evaluate the quantiles, we use the "New Algorithm" described here:
+
+  [MRL98] Manku, Rajagopalan & Lindsay, "Approximate Medians and other
+  Quantiles in One Pass and with Limited Memory", Proc. 1998 ACM SIGMOD,
+  Vol 27, No 2, p 426-435, June 1998.
+  http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.6.6513&rep=rep1
+  &type=pdf
+
+  The default error bound is (1 / N), though in practice the accuracy
+  tends to be much better.
+
+  Args:
+    num_quantiles: Number of quantiles to produce. It is the size of the final
+      output list, including the mininum and maximum value items.
+    buffer_size: The size of the buffers, corresponding to k in the referenced
+      paper.
+    num_buffers: The number of buffers, corresponding to b in the referenced
+      paper.
+    compare: (optional) Comparator function which is an implementation
+        of "a < b" taking two arguments (a and b). Which is later
+        converted to key function as Python 3 does not support `cmp`.
+    key: (optional) Key is a mapping of elements to a comparable key, similar
+      to the key argument of Python's sorting methods.
+    reverse: (optional) whether to order things smallest to largest, rather
+        than largest to smallest
+  """
+
+  # For alternating between biasing up and down in the above even weight
+  # collapse operation.
+  _offset_jitter = 0
+
+  # The cost (in time and space) to compute quantiles to a given accuracy is a
+  # function of the total number of elements in the data set. If an estimate is
+  # not known or specified, we use this as an upper bound. If this is too low,
+  # errors may exceed the requested tolerance; if too high, efficiency may be
+  # non-optimal. The impact is logarithmic with respect to this value, so this
+  # default should be fine for most uses.
+  _MAX_NUM_ELEMENTS = 1e9
+  _qs = None  # Refers to the _QuantileState
+
+  # TODO(mszb): For Python 3, remove compare and only keep key.
```

Comment:
Should we not introduce compare at all if this is only useful for python 2? Also from the use below, where we use cmp_to_key to do the mapping, is not py2 and py3 compatible?

/cc @tvalentyn to help with py3 questions.

Reply:
On Python 3 we should make the interface consistent with sorted() builtin on Python 3 from the beginning, so that we don't have to deprecate `compare` later. that is, we should allow only `key` and `reverse`, and drop `compare` on Python 3. This will be consistent with `Top()` combiners which underwent a similar change in https://github.com/apache/beam/pull/7842. Given that this is a new transform, I suggest we drop compare across the board, since it will ease migration to Python 3. 
For context, we plan to drop Python 2 support in Beam starting from 2020.

Reply:
So @aaltay, as @tvalentyn suggest, should drop the compare support?

Reply:
Let's drop compare support and go with @tvalentyn's suggestion.

Reply:
Done!